chainerx_cc/chainerx/cuda/cuda_device/linalg.cu

```diff
@@ -0,0 +1,129 @@
+#include "chainerx/cuda/cuda_device.h"
+
+#include <cstdint>
+#include <mutex>
+#include <type_traits>
+
+#include <cublas_v2.h>
+#include <cuda_runtime.h>
+#include <cusolverDn.h>
+#include <cuda_fp16.hpp>
+
+#include "chainerx/array.h"
+#include "chainerx/axes.h"
+#include "chainerx/backend.h"
+#include "chainerx/backend_util.h"
+#include "chainerx/cuda/cublas.h"
+#include "chainerx/cuda/cuda_runtime.h"
+#include "chainerx/cuda/cuda_set_device_scope.h"
+#include "chainerx/cuda/cusolver.h"
+#include "chainerx/cuda/data_type.cuh"
+#include "chainerx/cuda/float16.cuh"
+#include "chainerx/cuda/kernel_regist.h"
+#include "chainerx/device.h"
+#include "chainerx/dtype.h"
+#include "chainerx/error.h"
+#include "chainerx/float16.h"
+#include "chainerx/kernels/creation.h"
+#include "chainerx/kernels/linalg.h"
+#include "chainerx/kernels/misc.h"
+#include "chainerx/macro.h"
+#include "chainerx/native/native_device.h"
+#include "chainerx/routines/creation.h"
+#include "chainerx/routines/linalg.h"
+
+namespace chainerx {
+namespace cuda {
+
+class CudaSolveKernel : public SolveKernel {
+public:
+    void Call(const Array& a, const Array& b, const Array& out) override {
+        Device& device = a.device();
+        Dtype dtype = a.dtype();
+        CudaSetDeviceScope scope{device.index()};
+
+        CHAINERX_ASSERT(a.ndim() == 2);
+        CHAINERX_ASSERT(a.shape()[0] == a.shape()[1]);
+
+        auto solve_impl = [&](auto pt, auto getrf_bufferSize, auto getrf, auto getrs) {
```

Comment:
Can you make this a free function? Preferably in an unnamed namespace with a "Impl" suffix. You can also templetize it? Some reference https://github.com/chainer/chainer/blob/master/chainerx_cc/chainerx/cuda/cuda_device/reduction.cu

Reply:
Let's not get pulled to the CUDA naming convention with the camel case `getrf_bufferSize`.

Reply:
Done.